src/CryptographicHash.cpp

```diff
@@ -0,0 +1,163 @@
+// Copyright 2005-2017 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#include "murmur_pch.h"
+
+#include "CryptographicHash.h"
+
+class CryptographicHashPrivate {
+	public:
+		CryptographicHashPrivate(const EVP_MD *type);
+		~CryptographicHashPrivate();
+		void addData(const QByteArray &buf);
+		QByteArray result();
+
+	private:
+		EVP_MD_CTX *m_mdctx;
+
+		/// The result of the cryptographic hash.
+		/// This field is filled out by the result()
+		/// method.
+		/// Once this field is set, the m_mdctx is finalized
+		/// and can't be used anymore, so result() and addData()
+		/// must guard against that possibility.
+		QByteArray m_result;
+
+		CryptographicHashPrivate();
+
+		/// If m_mdctx is set, cleanupMdctx will destroy
+		/// the EVP_MD_CTX object and set m_mdctx to NULL.
+		void cleanupMdctx();
+};
+
+CryptographicHashPrivate::CryptographicHashPrivate()
+	: m_mdctx(NULL) {
+}
+
+void CryptographicHashPrivate::cleanupMdctx() {
+	if (m_mdctx) {
+		EVP_MD_CTX_destroy(m_mdctx);
+		m_mdctx = NULL;
+	}
+}
+
+CryptographicHashPrivate::CryptographicHashPrivate(const EVP_MD *type)
+	: m_mdctx(NULL) {
+
+	m_mdctx = EVP_MD_CTX_create();
+	if (m_mdctx == NULL) {
+		return;
+	}
+
+	int err = EVP_DigestInit_ex(m_mdctx, type, NULL);
+	if (err != 1) {
+		cleanupMdctx();
+		return;
+	}
+}
+
+CryptographicHashPrivate::~CryptographicHashPrivate() {
+	cleanupMdctx();
+}
+
+void CryptographicHashPrivate::addData(const QByteArray &buf) {
+	if (m_mdctx == NULL) {
+		return;
+	}
+
+	// If we have a result already, that means our m_mdctx
+	// object has been finalized, and can't be used anymore.
+	// In that case, return without adding the data. Our
+	// QCryptographicHash is "closed", so we can't add any
+	// more data.
+	if (!m_result.isNull()) {
+		return;
```

Comment:
Silently ignore invalid use?

I'd prefer this to fail hard, at least make m_result null to indicate an error/issue.